Additional items to heat conduction & diffusion physics

GiudGiud opened the pull request:
refs #28779  #28730 

with these additions that fit fine in the modules' Physics classes, I can derive new Physics downstream in TMAP and cover the entire assessment suite

Added ICs for usability

GiudGiud committed:
Add support for a functor heat  source in heat  conduction physics

GiudGiud committed:
Add a functor option to ADConvectiveHeatFluxBC

GiudGiud committed:
Add support for solid-fluid convection to heat conduction physics
Note that this will often only be used for far-field convection as the correlation for h imposes the use of the bulk temperatue

GiudGiud committed:
Better checks on diffusion BCs
Add support for setting ICs from diffusion physics
Allow override of MultiSpeciesDiffusion
refs #28730

GiudGiud committed:
Rename heat conduction physics base class to HeatConductionPhysicsBase
Add documentation for the base class

GiudGiud committed:
Add a test for convective BC + the functor heat source
refs #28779

GiudGiud committed:
Rename T_infinity to T_fluid for convective BC
refs #28779

GiudGiud committed:
Add more consts to physics-getters to be able to call routines in const contexts

GiudGiud committed:
Add initial conditions for multi-species diffusion
Add test for restart behavior

refs #28730

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2482835/), step Docs: sync website on 43c9ac5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28780/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2483216/), step Generate coverage on 0397d1e wanted to post the following:


## Framework coverage

Mutiple head shas were found when retrieving head reports:
  0397d1eb3cbd353a2d6ca5b1439f7135e919556d
  4d4ebd97f2e75a34591c3732456d547e708e08bb

## Modules coverage

Mutiple head shas were found when retrieving head reports:
  0397d1eb3cbd353a2d6ca5b1439f7135e919556d
  4d4ebd97f2e75a34591c3732456d547e708e08bb

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>9fb4f8</th>
      <th colspan=4>#28780 0397d1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.01%</td>
      <td>88.71%</td>
      <td>-0.30%</td>
      <td>82.89%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4275</td>
      <td>4351</td>
      <td>+76</td>
      <td>126</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>528</td>
      <td>554</td>
      <td>+26</td>
      <td>26</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28780/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28780/coverage/heat_transfer)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>9fb4f8</th>
      <th colspan=4>#28780 0397d1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.48%</td>
      <td>84.48%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>16999</td>
      <td>17002</td>
      <td>+3</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3123</td>
      <td>3123</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28780/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28780/coverage/navier_stokes)

### Scalar transport

<table>
  <thead>
    <tr>
      <th></th>
      <th>9fb4f8</th>
      <th colspan=4>#28780 0397d1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.44%</td>
      <td>88.86%</td>
      <td>+0.42%</td>
      <td>94.12%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>352</td>
      <td>367</td>
      <td>+15</td>
      <td>16</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>46</td>
      <td>46</td>
      <td>-</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28780/coverage_diff/scalar_transport)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28780/coverage/scalar_transport)

## Full coverage reports

<details>
<summary>Reports</summary>

- `framework` (FAILED)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/scalar_transport)
- `solid_mechanics` (FAILED)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/solid_properties)
- `stochastic_tools` (FAILED)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28780/coverage/xfem)

</details>

## Warnings

- `heat_transfer` new line coverage rate 82.89% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2483216/), step Verify coverage on 0397d1e wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)

lindsayad commented on modules/heat_transfer/include/bcs/ADConvectiveHeatFluxBC.h:
```diff
@@ -26,8 +26,14 @@ class ADConvectiveHeatFluxBC : public ADIntegratedBC
+  const ADMaterialProperty<Real> * const _htc;
+
+  /// Far-field temperature functor
+  const Moose::Functor<ADReal> * _T_infinity_functor;
```
```suggestion
  const Moose::Functor<ADReal> * const _T_infinity_functor;
```

lindsayad commented on modules/heat_transfer/include/bcs/ADConvectiveHeatFluxBC.h:
```diff
@@ -26,8 +26,14 @@ class ADConvectiveHeatFluxBC : public ADIntegratedBC
+  const Moose::Functor<ADReal> * _T_infinity_functor;
+
+  /// Convective heat transfer coefficient as a functor
+  const Moose::Functor<ADReal> * _htc_functor;
```
```suggestion
  const Moose::Functor<ADReal> * const _htc_functor;
```

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionCG.C:
```diff
@@ -62,6 +62,22 @@ HeatConductionCG::addFEKernels()
+    const std::string kernel_type = "BodyForce";
+    InputParameters params = getFactory().getValidParams(kernel_type);
+    params.set<NonlinearVariableName>("variable") = _temperature_name;
+    const auto functor_name = getParam<MooseFunctorName>("heat_source_functor");
```
```suggestion
    const auto & functor_name = getParam<MooseFunctorName>("heat_source_functor");
```

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFV.C:
```diff
@@ -73,6 +73,23 @@ HeatConductionFV::addFVKernels()
+    const std::string kernel_type = "FVBodyForce";
+    InputParameters params = getFactory().getValidParams(kernel_type);
+    params.set<NonlinearVariableName>("variable") = _temperature_name;
+    const auto functor_name = getParam<MooseFunctorName>("heat_source_functor");
```
```suggestion
    const auto & functor_name = getParam<MooseFunctorName>("heat_source_functor");
```

GiudGiud committed:
Address Alex's review

- avoid unnecessary copies using references
- keep pointers to functors for convective heat flux quantities const

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>
